### PR TITLE
Wip day11

### PR DESCRIPTION
**Pull request by mtsochacki**
It's day9 all over again.
I have a feeling if I did not use the "border of 9" solution it would be way cleaner.

### CODE REVIEW

File: 2021/day11/day11.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+public class day11 {
+    /*
+     * Creates a grid surrounded by a border of nines, e.g.:
+     * 999999
+     * 1234 912349
+     * 5678 --> 956789
+     * 9999 999999
+     * 999999
+     */
```
Comment (Lammatian) [inline 2021/day11/day11.java:13]: It's hard to parse this example. Mention that it creates a 2D array with the energy levels surrounded by a border. Also, in my opinion this would work nicer with a `static ArrayList<ArrayList<Point>> neighbours(int x, int y, int width, int height` method instead of the border.

File: 2021/day11/day11.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    public static ArrayList<ArrayList<Integer>> readInput(String filename) {
+        ArrayList<ArrayList<Integer>> input = new ArrayList<>();
```
Comment (Lammatian) [inline 2021/day11/day11.java:15]: `input` -> `octoGrid` or something like that

File: 2021/day11/day11.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        ArrayList<Integer> emptyLine = new ArrayList<>();
+        Scanner sc = null;
+        try {
+            sc = new Scanner(new File(filename));
+            sc.useDelimiter("");
```
Comment (Lammatian) [inline 2021/day11/day11.java:20]: is this necessary? Or is that the default?

File: 2021/day11/day11.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            input.add(emptyLine);
+            while (sc.hasNextLine()) {
+                ArrayList<Integer> line = new ArrayList<>();
```
Comment (Lammatian) [inline 2021/day11/day11.java:23]: `line` -> `octoRow` or something like that

File: 2021/day11/day11.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                line.add(9);
+                while (sc.hasNextInt()) {
+                    line.add(sc.nextInt());
+                }
+                line.add(9);
+                input.add(line);
+                if (sc.hasNextLine())
+                    sc.nextLine();
```
Comment (Lammatian) [inline 2021/day11/day11.java:31]: curly braces missing

File: 2021/day11/day11.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+        } catch (Exception e) {
+            System.out.println("Something went wrong " + e);
+        } finally {
+            sc.close();
+        }
+        return input;
+    }
+
+    public static void initialIncrease(ArrayList<ArrayList<Integer>> input) {
```
Comment (Lammatian) [inline 2021/day11/day11.java:48]: try to use verbs in the method names. I have no idea how to name this one though lol. Also `input` -> `octoGrid`

File: 2021/day11/day11.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+            for (int k = 1; k < input.get(j).size() - 1; k++) {
+                input.get(j).set(k, input.get(j).get(k) + 1);
+            }
+        }
+    }
+
+    public static void increaseOctopus(ArrayList<ArrayList<Integer>> input,
+            int x, int y) {
+        if (input.get(y).get(x) != 0)
+            input.get(y).set(x, input.get(y).get(x) + 1);
```
Comment (Lammatian) [inline 2021/day11/day11.java:59]: curly braces missing

File: 2021/day11/day11.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                if (input.get(j).get(k) > 9) {
+                    input.get(j).set(k, 0);
+                    increaseOctopus(input, k - 1, j);
+                    increaseOctopus(input, k + 1, j);
+                    increaseOctopus(input, k + 1, j - 1);
+                    increaseOctopus(input, k + 1, j + 1);
+                    increaseOctopus(input, k - 1, j - 1);
+                    increaseOctopus(input, k - 1, j + 1);
+                    increaseOctopus(input, k, j - 1);
+                    increaseOctopus(input, k, j + 1);
```
Comment (Lammatian) [inline 2021/day11/day11.java:75]: This should be a loop. You could do `for (Point p : neighbours(k, j, input.get(0).size(), input.size())) {...}` if you had the neighbours method

File: 2021/day11/day11.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        }
+    }
+
+    public static void increaseOctopus(ArrayList<ArrayList<Integer>> input,
+            int x, int y) {
+        if (input.get(y).get(x) != 0)
+            input.get(y).set(x, input.get(y).get(x) + 1);
+    }
+
+    public static Boolean processFlashes(ArrayList<ArrayList<Integer>> input) {
```
Comment (Lammatian) [inline 2021/day11/day11.java:62]: `input` -> `octoGrid`

File: 2021/day11/day11.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+                    increaseOctopus(input, k, j + 1);
+                }
+                if (input.get(j).get(k + 1) > 9
+                        || input.get(j + 1).get(k + 1) > 9
+                        || input.get(j + 1).get(k) > 9
+                        || input.get(j + 1).get(k - 1) > 9
+                        || input.get(j - 1).get(k + 1) > 9
+                        || input.get(j - 1).get(k) > 9
+                        || input.get(j - 1).get(k - 1) > 9
+                        || input.get(j).get(k - 1) > 9) {
```
Comment (Lammatian) [inline 2021/day11/day11.java:84]: Is this at all necessary? I don't see the return value of `processFlashes` ever being used. If it isn't remove it (and make this return `void`), if it is, use a for loop as above

File: 2021/day11/day11.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+                    explosionStatus = true;
+                }
+            }
+        }
+        return explosionStatus;
+    }
+
+    public static boolean isFlashReady(ArrayList<ArrayList<Integer>> input) {
```
Comment (Lammatian) [inline 2021/day11/day11.java:92]: `input` -> `octoGrid`

File: 2021/day11/day11.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+            for (int k = 1; k < input.get(j).size() - 1; k++) {
+                if (input.get(j).get(k) > 9) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public static int countFlashes(ArrayList<ArrayList<Integer>> input) {
```
Comment (Lammatian) [inline 2021/day11/day11.java:103]: `input` -> `octoGrid`

File: 2021/day11/day11.java [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+                if (input.get(j).get(k) != 0)
+                    return false;
+            }
+        }
+        return true;
+    }
+
+    public static void inspectOctopuses(int days) {
+        int flashCounter = 0;
+        boolean isAnyReadyToFlash = true;
```
Comment (Lammatian) [inline 2021/day11/day11.java:126]: I think this default value is unnecessary

File: 2021/day11/day11.java [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+        ArrayList<ArrayList<Integer>> input = readInput("data.txt");
```
Comment (Lammatian) [inline 2021/day11/day11.java:127]: `input` -> `octoGrid`

File: 2021/day11/day11.java [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        for (int i = 0; i < days; i++) {
+            initialIncrease(input);
+            isAnyReadyToFlash = true;
```
Comment (Lammatian) [inline 2021/day11/day11.java:130]: why `= true`? Why not `= isFlashReady(octoGrid)`?

File: 2021/day11/day11.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+                processFlashes(input);
+                isAnyReadyToFlash = isFlashReady(input);
+            }
+            flashCounter += countFlashes(input);
+            if (areAllFlashing(input)) {
+                System.out.println(i + 1);
+                break;
+            }
+        }
+        System.out.println(flashCounter);
```
Comment (Lammatian) [inline 2021/day11/day11.java:141]: I am not a fan of these `println`s. I would instead move the logic of processing a step to a new method, and then wrap this method in two separate methods for each part, e.g. `countFlashes` for part 1 and `findSynchronisationStep` for part 2. They should be simple, something like

```java
static int countFlashes(int days) {
  // read input
  int flashes = 0;
  for (int i = 0; i < days; ++i) {
    flashes += performOctoStep(octoGrid);
  }
  return flashes;
}
```
```

**Review by Lammatian — CHANGES_REQUESTED** [review]
Don't get discouraged by the amount of comments, many are a repetition of a single thing 😄

**Comment by mtsochacki** [comment]
Delimiter needs to stay (""), it allows to separate input digit by digit.
I think I need some more guidance when it comes to what exactly neighbours method is suppose to return.
I fixed everything else, let's hope it's more readable now.

**Comment by mtsochacki** [comment]
Also, for some reason I am unable to reply to your comments here, but I can do that in other PR

File: 2021/day11/day11.java [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+    public static int countAllFlashes(int step) {
+        ArrayList<ArrayList<Integer>> octoGrid = readInput("data.txt");
+        int flashes = 0;
+        for (int i = 0; i < step; i++) {
+            flashes += processStep(octoGrid);
+        }
+        return flashes;
+    }
+
+    public static int findSyncStep(int step) {
```
Comment (Lammatian) [inline 2021/day11/day11.java:126]: I would change `step` into `maxStep`, as this is the maximum step you're willing to check

File: 2021/day11/day11.java [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        int flashCounter = 0;
+        increaseAll(octoGrid);
+        while (isFlashReady(octoGrid)) {
+            processFlashes(octoGrid);
+        }
+        flashCounter += countFlashes(octoGrid);
+        return flashCounter;
+    }
+
+    public static int countAllFlashes(int step) {
```
Comment (Lammatian) [inline 2021/day11/day11.java:117]: I would change `step` to `numSteps` or just `steps`

**Review by Lammatian — CHANGES_REQUESTED** [review]
This looks very nice now, I love the `processStep`.

As for the neighbour function, there are two ways I see it being done. First is just enumerating all the possibilities (since there are only 4)

```java
ArrayList<Point> getNeighbours(int x, int y, int width, int height) {
  ArrayList<Point> neighbours = new ArrayList<>();
  if (x > 0) {
    neighbours.add(new Point(x - 1, y));
  }
  if (x < width - 1) {
    neighbours.add(new Point(x + 1, y));
  }
  if (y > 0) {
    neighbours.add(new Point(x, y - 1));
  }
  if (y < height - 1) {
    neighbours.add(new Point(x, y + 1));
  }
  return neighbours;
}
```

the second one is using for loops and is kind of neater, but possibly slightly less readable. This is definitely better if there are more neighbours (e.g. also diagonal) though

```java
bool inBounds(int x, int y, int width, int height) {
  return 0 <= x && x < width && 0 <= y && y < height;
}

ArrayList<Point> getNeighbours(int x, int y, int width, int height) {
  ArrayList<Point> neighbours = new ArrayList<>();
  int[] diffs = {-1, 1}; // this name is bad but just for the example

  for (int diff : diffs) {
    if (inBounds(x + diff, y)) {
      neighbours.add(x + diff, y);
    }
    if (inBounds(x, y + diff)) {
      neighbours.add(x, y + diff);
    }
  }
}
```

File: 2021/day11/day11.java [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        }
+        if (y > 0) {
+            neighbours.add(new Point(x, y - 1));
+        }
+        if (y < height - 1) {
+            neighbours.add(new Point(x, y + 1));
+        }
+        return neighbours;
+    }
```
Comment (Lammatian) [inline 2021/day11/day11.java:66]: This is too long and too repetitive - if it was only 4 neighbours as I said in an earlier comment, I would accept such a solution, but with 8 it's too long. You can either, as I said before, write an `isInBounds` method or you can even implicitly fit it in here (don't include the comments ofc):

```java
for (int nx = Math.max(0, x - 1); nx < Math.min(width, x + 2); ++nx) {  // Implicitly checking the bounds for x
  for (int ny = Math.max(0, y - 1); ny < Math.min(height, y + 2); ++ny) {  // Implicitly checking the bounds for y
    if (nx != x || ny != y) {  // Omit the point itself
      neighbours.add(new Point(nx, ny));
    }
  }
}
```

**Review by Lammatian — CHANGES_REQUESTED** [review]
Neighbours adjustment + minor comments

**Review by Lammatian — APPROVED** [review]
Lovely, great work 👍🏼
